File: composeApp/src/androidMain/kotlin/cz/miroslavpasek/pigeonnavigator/ui/home/model/HudFormatting.kt
package cz.miroslavpasek.pigeonnavigator.ui.home.model

import cz.miroslavpasek.pigeonnavigator.domain.settings.AltitudeUnit
import cz.miroslavpasek.pigeonnavigator.domain.settings.SpeedUnit
import kotlin.math.roundToInt

/**
 * Display-side conversions for altitude and speed used by the HUD cluster.
 *
 * Lives in the home UI layer so the domain enums stay free of formatting
 * concerns.
 */
internal fun Double.toDisplayAltitude(unit: AltitudeUnit): Int = when (unit) {
    AltitudeUnit.Feet -> (this * 3.280839895).roundToInt()
    AltitudeUnit.Meters -> roundToInt()
}

internal fun Float.toDisplaySpeed(unit: SpeedUnit): Int = toDouble().toDisplaySpeed(unit)

internal fun Double.toDisplaySpeed(unit: SpeedUnit): Int = when (unit) {
    SpeedUnit.Knots -> (this * 1.943844492).roundToInt()
    SpeedUnit.KilometersPerHour -> (this * 3.6).roundToInt()
    SpeedUnit.MilesPerHour -> (this * 2.236936292).roundToInt()
    SpeedUnit.MetersPerSecond -> roundToInt()
}

internal fun AltitudeUnit.displayLabel(): String = when (this) {
    AltitudeUnit.Feet -> "ft"
    AltitudeUnit.Meters -> "m"
}

internal fun SpeedUnit.displayLabel(): String = when (this) {
    SpeedUnit.Knots -> "kt"
    SpeedUnit.KilometersPerHour -> "km/h"
    SpeedUnit.MilesPerHour -> "mph"
    SpeedUnit.MetersPerSecond -> "m/s"
}

File: iosApp/iosApp/UI/Home/Model/HudFormatting.swift
import Foundation
import Shared

/// Display-side conversions for altitude and speed used by the home HUD.
///
/// Mirrors `composeApp/.../home/model/HudFormatting.kt` so the Android and
/// iOS HUDs render identical numbers.
enum HudFormatting {
    static func displayAltitude(meters: Double, unit: DomainAltitudeUnit) -> Int {
        switch unit {
        case .feet:
            return Int((meters * 3.280839895).rounded())
        case .meters:
            return Int(meters.rounded())
        default:
            return Int(meters.rounded())
        }
    }

    static func displaySpeed(metersPerSecond: Double, unit: DomainSpeedUnit) -> Int {
        switch unit {
        case .knots:
            return Int((metersPerSecond * 1.943844492).rounded())
        case .kilometersperhour:
            return Int((metersPerSecond * 3.6).rounded())
        case .milesperhour:
            return Int((metersPerSecond * 2.236936292).rounded())
        case .meterspersecond:
            return Int(metersPerSecond.rounded())
        default:
            return Int((metersPerSecond * 3.6).rounded())
        }
    }

    static func altitudeUnitLabel(_ unit: DomainAltitudeUnit) -> String {
        switch unit {
        case .feet:
            return "ft"
        case .meters:
            return "m"
        default:
            return "m"
        }
    }

    static func speedUnitLabel(_ unit: DomainSpeedUnit) -> String {
        switch unit {
        case .knots:
            return "kt"
        case .kilometersperhour:
            return "km/h"
        case .milesperhour:
            return "mph"
        case .meterspersecond:
            return "m/s"
        default:
            return "km/h"
        }
    }
}
